### ISSUE TITLE
Elm 0.19, OS X/Ubuntu: Compiler crashes with "not enough bytes", on large project after 1) failed build, 2) trying to build again

### ISSUE THREAD

**Issue by wmakley** [elm/compiler#1837]
Affects OS X and Linux consistently.

- Elm 0.19
- OS X Mojave
- Ubuntu 18.04

**Steps to reproduce:**

1. Create a large project.
2. Introduce a build-failing error into the code. 
3. Compile the first time.
4. Compile a second time.

**Workaround:**

Delete "elm-stuff" and try again.

**Example output:**

```sh
make MyProject.js
elm make  src/MyProject.elm --output ./release/MyProject.js
Dependencies loaded from local cache.
Dependencies ready!
Detected errors in 1 module.
-- TYPE MISMATCH --------------------------------------- src/MyProject.elm

The 1st argument to `initWithFlags` is not what I expect:
# [...]

# Try again!
elm make  src/MyProject.elm --output ./release/MyProject.js
elm: not enough bytes
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary.hs:212:21 in binary-0.8.5.1:Data.Binary
make[1]: *** [MyProject.js] Error 1
make: *** [test] Error 2
```

**Comments**

I have noticed that it seems to occur more often if the original compiler error is very "deep" in the app. I have a very complicated widget that I re-use repeatedly very deep in most of my views, and introducing an error into it almost always causes this bug.

This is really messing up my workflow, because the workaround makes builds take much longer. I have several codebases that have grown to be affected by this bug. The main issue with submitting a quality bug report is that all my affected code is proprietary, so I fully understand if you want to moderate this away as a "low quality" bug report. I would love to send the maintainers my codebase, I just can't post it on github.

Thanks!

**Comment by purcell** [comment]
I see this constantly too, unpredictably in a large Elm codebase with several errors present: I'm working on upgrading from 0.18 to 0.19. Haven't figured out a pattern yet.

**Comment by mapmarkus** [comment]
I'm seeing this too, more often now that a project I'm working on is growing really big.

I'm using Ubuntu (with Docker over macOS) and a file watcher to recompile whenever any file changes. The error appears randomly, sometimes when I change a bunch of files at the same time, sometimes when I change a single file.

The workaround is indeed removing `elm-stuff`.

**Comment by mbark** [comment]
Since the problem for us seems to occur from having only a `.elmi` file without a matching `.elmo` one we run a script to ensure consistency between `.elmi` and `.elmo` files before running `elm make`.

So in our `gulpfile.js` we have a task that runs the function below before doing `elm make` or `elm bundle`.
```
return function() {
    var elmStuff = src + "elm-stuff";
    function checkFiles(fileToCheck, matchingFileType) {
        var files = glob.sync(elmStuff + "/**/*." + fileToCheck, {});

        for (var i = 0; i < files.length; i++) {
            var file = files[i];
            var matchingFile = file.replace(/\.[^/.]+$/, "") + "." + matchingFileType;

            if(fs.existsSync(file) && !fs.existsSync(matchingFile)) {
                console.error('Found ' + file + ' but not expected matching file ' + matchingFile);
                rimraf.sync(file);
            }
        }
    }

    checkFiles('elmi', 'elmo');
    checkFiles('elmo', 'elmi');
};
```

The script just goes through the `elm-stuff` directory and checks for all `.elmi` files if there is no matching `.elmo` one, if so it removes the `.elmi` file and then vice versa for all `.elmo` files. This does means the entire `elm-stuff` directory doesn't have to rebuilt at least.

**Comment by evancz** [comment]
For people seeing this issue, are you using an editor that is running `elm make` automatically? Is it possible that this happens when there are two concurrent `elm make` commands issued?

**Comment by wmakley** [comment]
I wish, but I have a repo that it happens on without any editor open 100% of the time.

**Comment by popara** [comment]
@evancz  
I am using webpack with a elm-loader and normal settings. I have even tried to run webpack's `elm make` command from terminal. Same thing.

**Comment by mapmarkus** [comment]
@evancz 
You actually may have a point, I can confirm that triggering many compilation events causes two types of errors:

1. `The binary data at /project/elm-stuff/0.19.0/name-of-file.elmo is corrupt`. Removing this file solves the problem.

2. `elm: not enough bytes`. The problem described in this issue.

The events were fired by saving a file several times in a row (3-4 usually). I used a file that is imported by many other modules, so the compilation takes a little longer. At least one of the errors happen every time after the third or fourth compilation. 

I use a file watcher (chokidar-cli) that triggers `elm make` if a file changes. The tool has `--debounce` and `--throttle` parameters, but I use with the default values, that are `400` and `0` respectively.

**Comment by mapmarkus** [comment]
I've been playing with a simple temporary solution that apparently works in my particular case. Instead of running `elm make` directly, the file watcher runs this script:

```bash
if [[ -a .compiling ]]; then
    echo "Still compiling..."
else
    touch .compiling
    elm make src/Main.elm --output /path/to/public/elm.js
    rm .compiling
fi
```

This works because `elm-make` blocks, and `rm .compiling` is not executed until it finishes.

Maybe the solution is as simple as saving some _flag_ (like a temporary file or something smarter and more efficient) and then blocking any subsequent compilation until the current one finishes.

**Comment by mbark** [comment]
@evancz I can reproduce the error by removing `elm-stuff` (to ensure all modules have to be rebuilt), editing a file to make it produce an error, running `elm make src/**/*.elm` (which will fail to compile due to the error) and then `elm make src/**/*.elm` again which will print the `elm: not enough bytes` error.

So in our case the problem doesn't seem to be related to having concurrent `elm make` commands. Looking at the files it seems as if it might be that the compiler encounters the error and then finishes in a half-completed state leaving an `.elmi` file that isn't fully built and no matching `.elmo` file at all.

**Comment by martenbr** [comment]
As @mbark describes above the problem happens when invoking a simple elm make command twice. An interesting observation however is that the size of the .elmi file does vary between runs:

`-rw-rw-r--. 1 martenb martenb 953846 Nov 19 09:52 elm-stuff/0.19.0/Widget-RecTable.elmi`
`-rw-rw-r--. 1 martenb martenb 1084847 Nov 19 09:51 elm-stuff/0.19.0/Widget-RecTable.elmi`
`-rw-rw-r--. 1 martenb martenb 691839 Nov 19 09:52 elm-stuff/0.19.0/Widget-RecTable.elmi`

This indicates a concurrency issue so i did a strace and saw the following:
- The compiler seems to compile things in parallel (at the very least it is writing the .elmi in parallel)
- The compiler keeps writing to the broken elmi file even when it start printing the compilation error.
- The compiler exits when it has finished printing the compilation error, it keeps writing to the elmi file until this point but it never calls close on it.
- Printing the compilation error is a relatively slow operation

In other words it seems that to trigger this issue you need to have a very big file (that compiles fine) which the compiler is not able to finish writing during the time it is printing the compilation error. This would explain why only large projects are affected.

**Comment by xarvh** [comment]
Having the same issue on a medium-sized project
```
elm make src/Main.elm --output build/app.js
elm: not enough bytes
CallStack (from HasCallStack):
  error, called at src/Data/Binary.hs:212:21 in binary-0.8.6.0-1SWOkUSZj1B1k0FQ5fyFqI:Data.Binary
make: *** [Makefile:3: default] Error 1
```

The problem happens on one of my computers but not on the other, which seems consistent with @martenbr 's explanation.

**Comment by AleXoundOS** [comment]
Interesting, will the issue go away if restrict the number of available cores to the elm compiler process?

**Comment by Erudition** [comment]
Started to get this too, I'm using Elmjutsu on Atom. So far removing `elm-stuff/` works.

**Comment by n1k0** [comment]
Interestingly enough, a VSCode user has found that disabling live linting made the issue [going away](https://twitter.com/IlliaKalaur/status/1135656492420517889):

![](https://i.imgur.com/IEw9XTI.png)

I wonder if this issue in Elmjutsu has the same root cause, and if so what could be achieved to prevent it to happen.

**Comment by jubarbie** [comment]
I had the same issue when my elm projects were on an external hard drive. Working locally fixed it.
It is a bit annoying though since I'm working on hundreds of projects at the same time.

**Comment by evancz** [comment]
@wmakley, does this problem persist with 0.19.1?

**Comment by evancz** [comment]
It looks like a bunch of different issues have been reported in this thread.

The issues related to running `elm make` multiple times at once on the same project should be resolved by 0.19.1. I think the scenario explained by @martenbr can still happen though. I am going to start a new issue that specifically highlights that possibility.

If particular individuals are still seeing this kind of thing with 0.19.1, please open an issue of your own. **Do not add on to an existing issue.** Please share as much information about your development environment as possible. Editor? File watcher? OS? Etc.

**Comment by wmakley** [comment]
> @wmakley, does this problem persist with 0.19.1?

Nope! Miraculously I found my old repo+tag and it still had the same issue in 0.19. Then I installed 0.19 and ran into many new compiler errors and tried to work through some of them and run make multiple times, but did not encounter this issue again.

**Comment by toastal** [comment]
I'm having issues with running Elm v0.19.1 in parallel via `lerna` + `webpack` in certain environments. We moved to running in sequence to resolve our build issue in Heroku (although it seems to be fine locally on everyone's laptops).

**Comment by evancz** [comment]
@toastal, can you open a new issue with an SSCCE so we can reproduce and fix if the root problem is in the `elm` binary?
